storage/storage.c

```diff
@@ -126,6 +126,12 @@ const uint32_t V0_PIN_EMPTY = 1;
 // The length of the ChaCha20 IV (aka nonce) in bytes as per RFC 7539.
 #define CHACHA20_IV_SIZE 12
 
+#ifdef FLASH_BYTE_ACCESS
+#define CHACHA20_IV_PADDING 0
+#else
+#define CHACHA20_IV_PADDING 4
```

Comment:
Due to encrypted data are organized, the buffering approach does not work (unless the buffer is same size as the as the whole item size, but then using `update_bytes` is useless + we don't have any sort of dynamic allocation on this level so its problematic).

The reason is that the `TAG` is stored after the `IV` at the beginning, but is computed and written after the encrypted data. Therefore, the IV and (beginning of) the message would have to be stored until the tag is computed. In more general situation, we cannot tell anything about the order of data fed into update_bytes, therefore, we would have to buffer the whole thing.

One way out is to change the update_bytes function so that it does not accept offset as a parameter, but collects the data sequentially: but that require storing the tag after the encrypted data, which is incompatible with currently stored data. Which leads to upgrading the storage, or keeping two versions of encrypted data. 

Also this would limit the functionality of the `update_bytes` function. 

This is to some extent related to the comment above related to the data structure of full-size items above. If we add the 14 bytes of data to the same quadword as key, we need to buffer the key until the data is known.

Reply:
Good catch. I forgot that we put the tag at the beginning. That wasn't a very good choice.
> One way out is to change the update_bytes function so that it does not accept offset as a parameter, but collects the data sequentially

That's what I assumed we'd do, but as you correctly note, it's incompatible with currently stored data.